Create mandate for subscriptions which do not have a mandate created previously.

Mayisha opened the pull request:
Fixes #2607 

The changes in this PR fix a few issues related to mandate in a subscription and renewal. The main target of this PR is to fix the error that happens while processing a `0` amount subscription which needs a mandate. A mandate is required only for the cards issued in India.

The main issues handled in this PR are-

1. Unable to process 0/month subscriptions. We [add](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/compat/trait-wc-stripe-subscriptions.php#L598-L612) the `mandate` data to the request and on Stripe end it creates a mandate if required by attaching a payment method to it. The amount needs to be a positive integer and so it fails when the amount is `0`.
2. If we just bypass the mandate options for `0` amount subscriptions and later update the subscription amount to a positive value to be charged, we need to create a mandate and save it for later. 
3. If we change the subscription amount or the payment method, the existing mandate won't work and we need to create a new mandate for the amount and/or the payment method.
4. When UPE is enabled, if we buy a subscription with a saved card it does not create a mandate. 
5. When UPE is disabled, if we buy a subscription with a saved card it creates a mandate and use the same mandate in renewal, But the manual renewal fails.


## Changes proposed in this Pull Request:
 - avoid setting mandate options when the subscription amount is 0. (fixes 1)
 - add mandate data to renewal if not present in any previous renewal orders. (fixes 2)
 - Save mandate id in the order meta, in case it was created in some later renewals, and reuse it on the later renewals. (fixes 2)
 - If the saved mandate is for a different amount and/or different payment method, create a mandate again and save it to the current order meta (fixes 3)
 - Pass `confirm` and `setup_future_usage` attributes correctly when an existing `mandate` is passed with the request. (fixes 4, 5) https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2633/commits/947acf13125292501761682cb945426d6721d535 https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2633/commits/b8d987dc8bde561942211c73deea33d7f76b5b64

### Related resources for reading
- https://stripe.com/docs/invoicing/india-emandate-guide
- https://stripe.com/docs/india-recurring-payments?integration=subscriptions

## Testing instructions
### Prerequisite 
- Change the store currency to INR
- Make sure the webhooks are working

### Testing 0 amount subscription (Test with both UPE enabled and disabled)
- Create a subscription with 0 amount and a sign-up fee.
- As a shopper go to the shop page and add the subscription to your cart.
- Use card `4000003560000123` to complete the purchase.
- Make sure that there is no error and the subscription is active.
- Check logs in your [Stripe dashboard](https://dashboard.stripe.com/test/logs?method[0]=post&method[1]=delete&direction[0]=connect_in&direction[1]=self) also to ensure there is no error.
- Find Id of the order, check your DB and make sure that there is no `_stripe_mandate_id` meta set for the order.
- In admin dashboard, go to the subscription edit page and edit the subscription amount to a positive value (500/1000/5000 etc)
- Go to the `my account` page as the shopper and renew the subscription manually.
- Complete the renewal and make sure there is no error.
- Check your DB and make sure that there is now a `_stripe_mandate_id` meta set for the current order.
- Manually renew the order again and check your logs (in Stripe dashboard or in WooCommerce -> Status -> Logs). Make sure that the mandate saved in the previous order is sent with the request.

### Testing saved card with UPE enabled
- Disable UPE
- Create a subscription product with a positive amount
- As a shopper, go to the shop page as a logged in user, make sure you have the `4000003560000123` indian card saved in the payment methods and select the subscription product.
- Complete the purchase
- Make sure a mandate is created. Check your DB and make sure that there is a `_stripe_mandate_id` meta set for this order.
- Manually renew the order again and check your logs (in Stripe dashboard or in WooCommerce -> Status -> Logs). Make sure that the mandate saved in the previous order is sent with the request.

### Testing saved card with UPE disabled
- Enable UPE
- Create a subscription product with a positive amount
- As a shopper, go to the shop page as a logged in user, make sure you have the `4000003560000123` indian card saved in the payment methods and select the subscription product.
- Complete the purchase
- Make sure a mandate is created. Check your DB and make sure that there is a `_stripe_mandate_id` meta set for this order.
- Manually renew the order again and check your logs (in Stripe dashboard or in WooCommerce -> Status -> Logs). Make sure that the mandate saved in the previous order is sent with the request.

### ~~Testing with a different amount~~ (Skip this section)
- ~~Create a subscription with a positive value (500/1000/2000 etc)~~
- ~~As a shopper go to the shop page and add the subscription to your cart.~~
- ~~Use card `4000003560000123` to complete the purchase.~~
- ~~Make sure that there is no error and the subscription is active.~~
- ~~Find Id of the order, check your DB and make sure that there is a `_stripe_mandate_id` meta set for the order.~~
- ~~Renew this order and check the request body in the logs (in Stripe dashboard or in WooCommerce -> Status -> Logs) to ensure that the mandate created in the previous order is sent with the request.~~
- ~~Find Id of the order, check your DB and make sure that there is no `_stripe_mandate_id` meta set for this renewal order.~~
-  ~~In admin dashboard, go to the subscription edit page and edit the subscription amount to a different positive value (1500/3000/5000 etc)~~
- ~~Go to the `my account` page as the shopper and renew the subscription manually.~~
- ~~Find Id of the order, check your DB and make sure that there is a `_stripe_mandate_id` meta set for the order with a different mandate id than before.~~


### Testing critical flows
Make sure all [critical flows](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows) work as intended with both the Classic and UPE checkouts. Make sure to test both Shortcode and Block-based flows. Test with a normal card, 3DS card and Indian card `4000003560000123`

### Testing automatic renewal
Follow and test the steps mentioned in the [original PR](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2561#issue-1606058756) of implementing e-mandate.

---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

Mayisha committed:
fix logic for mandate
    - avoid setting mandate options when subscription amount is 0
    - add mandate data to renewal if not present in parent order

reykjalin commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -570,9 +570,21 @@ public function add_subscription_information_to_intent( $request, $order, $prepa
 			$parent_order    = wc_get_order( $parent_order_id );
 
 			if ( $parent_order ) {
-				$mandate = $parent_order->get_meta( '_stripe_mandate_id', true );
+				$mandate        = $parent_order->get_meta( '_stripe_mandate_id', true );
+				$renewal_amount = WC_Stripe_Helper::get_stripe_amount( $renewal_order->get_total() );
+
 				if ( ! empty( $mandate ) ) {
 					$request['mandate'] = $mandate;
+					return $request;
+				} elseif ( 0 !== $renewal_amount ) {
+					$request['payment_method_options']['card']['mandate_options']['amount_type']     = 'fixed';
+					$request['payment_method_options']['card']['mandate_options']['interval']        = $renewal_order->get_billing_period();
+					$request['payment_method_options']['card']['mandate_options']['interval_count']  = $renewal_order->get_billing_interval();
+					$request['payment_method_options']['card']['mandate_options']['amount']          = $renewal_amount;
+					$request['payment_method_options']['card']['mandate_options']['reference']       = $order->get_id();
+					$request['payment_method_options']['card']['mandate_options']['start_date']      = $renewal_order->get_time( 'start' );
+					$request['payment_method_options']['card']['mandate_options']['supported_types'] = [ 'india' ];
+
 					return $request;
 				}
```
The entire point of this was to just see if the mandate existed on the parent order, so I don't think this is the right place to add this logic. My initial implementation would've been better served by extracting this into a function so the code would've looked like:

```php
// instead of if ( 1 === count( $renewasl ) ) { /* ... */

$renewals = wcs_get_subscriptions_for_renewal_order( $order );
$mandate  = $this->get_mandate_from_parent_order( $order );
if ( $mandate ) {
  $request['mandate'] = $mandate;
  return $request;
}
```

All of that to say: I don't think the fix is in the right place here 🙂

Mayisha replied:
@reykjalin thanks for your review and guideline. I have made some changes to incorporate some of the suggestions you made [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2633#pullrequestreview-1464268566) like saving the mandate in the parent order.

I am a bit confused about this suggestion. What would be the best place for these changes? :thinking: 
I added the changes here to support the following scenario-
The subscription initially has 0 amount. So when the user first purchases the sub, there is no mandate. Then the subscription value is updated and now it has a positive value to charge which requires a mandate. At this point when a renewal order comes there is no mandate in the parent but we must send these `mandate_options` with the request. As this is a renewal order, without this else section it returns from [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/compat/trait-wc-stripe-subscriptions.php#L587) without adding the `mandate_options`.

reykjalin replied:
All I meant to say was that these changes should exist outside the `if ( 1 === count( $renewal_orders ) )` statement, something like:

```php
// Check if mandate already exists.
if ( 1 === count( $renewal_orders ) ) {
    // ...
    if ( $parent_order ) {
        // ...
    }
}

// Make sure we handle renewal orders that don't have mandates.
if ( 0 !== $renewal_amount ) {
    // Continue your code here.
}
```


I suggest this because the section you initially added the code to is intended to only check if there exists a mandate on the parent order, and combining that with handling free trials (or $0 subscriptions) would be confusing.

***

I honestly think the best way to handle this is probably to do one of the following:

1. Check what the renewal amount will be when a $0 sub is processed, create a mandate with that correct amount, save that to the parent order, and use that for renewal payments — this would work with the current "check the parent order for a mandate" approach; or
2. Create the mandate on the first renewal order that has a non-zero amount (what you're doing now), and either:
    1. Save the mandate to the parent order so it can be used for future purchases (this would work with the current "check the parent order for a mandate" approach); or
    2. Save the mandate to the renewal order. This would require changing the logic here such that we don't only check the parent order for a mandate but also previous renewal orders.

***

And if that all wasn't complicated enough you can create subscriptions without a parent order, in which case (1) and (2i) wouldn't work, so (2ii) is probably the best approach? But that complexity should probably be handled in a separate PR.

Mayisha replied:
Thanks @reykjalin for the suggestions. I had I have implemented 2i now with a few additional checks before reusing the mandate. :tada: 
I would appreciate your re-review here in case I have missed any other corner cases (like the subscriptions without a parent order case). Let me know if you can think of any possible regression too.

reykjalin replied:
You've addressed the main point here (not merging "finding the mandate" and "handle 0 cost subscriptions") so I'll mark this as resolved. Danae has hit upon some great points in her review and I added a few other comments of my own to continue the related discussions 🙂 

Great work! 👏

reykjalin reviewed (changes requested):
I think the fix is fine (didn't test the code), but I don't think it's in the right place (see comment). I'm also unsure if it's complete 🤔 

The tricky part here is that, with the current implementation (both before and after this PR), if the parent order is a $0 order -- or if there is no parent order -- we're always creating a new mandate because we either (a) never created a mandate and saved it to the parent order; or (b) don't have a parent order in the first place.

So along with the fix that allows for $0 orders we _also_ need to include a fix that can look up a mandate ID from previous orders. The recurring payments for a subscription all need to use the same mandate; if you create a new one the customer has to go through the "sign-up" process for the subscription all over again instead of just confirming the renewal payment.

I'm not exactly sure what the best way to achieve this is. Maybe to always save the mandate on the first order in the subscription, even if the mandate was generated on a different order (e.g. $0 payment first, renewal has an actual payment)? Maybe we save the mandate ID on the order where a payment was first made and then every subsequent order as well so we only have to check the previous renewal?

Honestly, this is probably going to need some exploration and I'm happy to provide more advice or discuss more in-depth either here or in Slack 🙂

Mayisha committed:
set `setup_future_usage` as `off_session` for subscription payments with UPE enabled

Mayisha committed:
pass `confirm` and  `setup_future_usage` attributes correctly with `mandate`
- Without `confirm=true` and with `setup_future_usage` set, passing mandate does not work

Mayisha committed:
add mandate to order
- for one time order, add mandate to the current order
- for subscription, add mandate to the parent order

Mayisha committed:
Merge branch 'develop' into fix/2607-0-sub

Mayisha committed:
Merge branch 'develop' into fix/2607-0-sub

Mayisha committed:
save mandate id to current order meta instead of parent order

Mayisha committed:
find mandate from a previous renewal order

Mayisha committed:
compare parent order amount without the signup fee

a-danae reviewed (commented):
Thank you very much for working on this, Mayisha! 

I tested these scenarios that had errors before and worked fine:
✅ Purchase a 0-subscription with a sign-up fee. EUR. UPE
✅ Purchase a 0-subscription with a sign-up fee. EUR. Non-UPE

I also went through almost all the testing steps from the PR description:
✅ Testing 0 amount subscription. INR. Non-UPE
✅ Testing saved card with UPE enabled
✅ Testing saved card with UPE disabled
✅ Testing with a different amount

❓ Testing 0 amount subscription. INR. UPE
This flow worked fine with the suggested steps. But I had a problem using a new payment method instead of a saved one for the manual renewal. These are the steps:
- In the checkout page, select “Use a new payment method”
- Entered `4000 0035 6000 0123`
- Clicked on “Renew subscription”
- An error message shows up saying 
  “stripe.confirmSetup(): the \`confirmParams.return_url\` argument is required unless passing \`redirect: 'if_required'\`"
There's an entry in the debug.log file saying 
  `PHP Notice: Undefined index: payment_method in /var/www/html/wp-content/plugins/woocommerce-gateway-stripe/includes/compat/trait-wc-stripe-subscriptions.php on line 573`
- It works fine when using the saved card. I couldn't replicate this with non-UPE.

Mind checking if you get the same behavior?

I have the following pending, but I wanted to leave the feedback I got in the meantime.
- [x]  Testing critical flows
- [ ]  Testing automatic renewal
- [x] Checking out the code

a-danae commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -559,22 +559,42 @@ public function add_subscription_information_to_intent( $request, $order, $prepa
 		// If we've already created a mandate for this order; use that.
 		$mandate = $order->get_meta( '_stripe_mandate_id', true );
 		if ( isset( $request['confirm'] ) && filter_var( $request['confirm'], FILTER_VALIDATE_BOOL ) && ! empty( $mandate ) ) {
+			$request['confirm'] = 'true';
```
Having `$request['confirm']` set to a truthy value is a requirement (because of `filter_var( $request['confirm'], FILTER_VALIDATE_BOOL )` in line 561) for reaching this point. Why would we need to set it again?

Mayisha replied:
It was a copy-paste error from my end. :grimacing:  
Removing this line.

a-danae commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -611,9 +635,51 @@ public function add_subscription_information_to_intent( $request, $order, $prepa
+			// and for a specific amount in Stripe so the payment method and amount should also match to reuse the mandate.
+			if ( ! empty( $mandate ) && $renewal_order_payment_method === $payment_method ) {
+				// phpcs:ignore
+				if ( $renewal_order_amount == $order_amount ) {
```
Would it be better to cast these variables than to use a loose comparison?

Mayisha replied:
Removed this comparison as discussed [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2633#discussion_r1264767126)

a-danae commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -611,9 +635,51 @@ public function add_subscription_information_to_intent( $request, $order, $prepa
+			// and for a specific amount in Stripe so the payment method and amount should also match to reuse the mandate.
+			if ( ! empty( $mandate ) && $renewal_order_payment_method === $payment_method ) {
+				// phpcs:ignore
+				if ( $renewal_order_amount == $order_amount ) {
```
An edgy scenario where this check passes but the payment fails because of a mandate amount mismatch:

- Enable taxes → Set “Prices entered with tax” to “Yes, I will enter prices inclusive of tax”
- Set a standard rate of 20%
- Under WooCommerce → Subscription, click on “Add Subscription”
- Assign the subscription to a customer for which the tax rate applies, set the status to “Active”
- Add a subscription item, but don’t click on “Recalculate” or include the taxes ![z7iIX2.png](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/64006fad-8f33-4ba9-bbe1-b2d384d134ba)
- Click on “Publish”
- As the customer, go to My Account → Subscriptions → this new subscription → Renew now
- Select an already saved 3Ds card, like 4000003560000123, and click on “Renew subscription”
- As a merchant, go edit the subscription you created earlier
- Click on “Recalculate” to include the taxes for the merchant and save
- As the customer, renew this subscription with the same card you used before
- Notice that:
    - You can’t complete the renewal because you get an error message saying “Unable to process this payment, please try again or use alternative method.”
    - In the WC → Stripe logs, you get
    
    ```
    Error: stdClass Object
    (
        [error] => stdClass Object
            (
                [code] => payment_intent_mandate_invalid
                [doc_url] => https://stripe.com/docs/error-codes/payment-intent-mandate-invalid
                [message] => This mandate can only be used for a payment of 250000. You attempted a payment of 300000.
                [request_log_url] => https://dashboard.stripe.com/test/logs/req_xxxxxx?t=1689442960
                [type] => invalid_request_error
            )
    
    )
    ```


**There’s another less edgy way to replicate this:**

- Set two tax standard rates: 20% for California, 0% for New York
- Create the subscription for the customer using a shipping address from New York
- As the customer, renew the subscription using the shipping address from New York
- Go renew the subscription again using the same card, but using a shipping address from California
- Notice you get the same error mentioned before

a-danae replied:
When there's a mismatch between the subscription and renewal totals, a new mandate is created for each subsequent renewal. Not sure if this is expected. Steps:

- Set two tax standard rates: 20% for California, 0% for New York
- Create the subscription for the customer using a shipping address from New York
- As the customer, renew the subscription using a shipping address _from California_
- Go renew the subscription using a saved card
- Note the mandate in WC logs -> Stripe logs
- Renew the subscription again using the same shipping address and the same card
- Notice the mandate differs from one in the last renewal, even though the total and payment method are the same. I'd expect the previously created mandate to be reused

a-danae commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -559,22 +559,42 @@ public function add_subscription_information_to_intent( $request, $order, $prepa
 			return $request;
 		}
 
 		$renewals = wcs_get_subscriptions_for_renewal_order( $order );
```
[Shouldn't](https://github.com/Automattic/woocommerce-subscriptions-core/blob/0303a5a84b217114c122bef3c62cf6cfa8fdddad/includes/wcs-renewal-functions.php#L110-L118) this be `$subscriptions`, instead of `$renewals`? Or another name, since `$subscriptions` is used below.

I know this wasn't introduced in this PR, but it may be worth adjusting it since we're working with that variable.

Mayisha replied:
renamed in [3476211](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2633/commits/34762119bd4ddbc032c841b6747894625f7505f9)

a-danae commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -559,22 +559,42 @@ public function add_subscription_information_to_intent( $request, $order, $prepa
+			}
+		}
+
+		// Add mandate options to request to create new mandate if mandate id does not already exist in a previous renewal.
```
Just a note that we wouldn't be checking previous renewals but the actual subscription. 
Related to my comment in line 566/588 about `wcs_get_subscriptions_for_renewal_order`

Mayisha replied:
For finding the mandate, we are actually checking the previous renewal order and the parent order. As the mandate is saved in the order's meta, not in the subscription itself.

a-danae reviewed (commented):
✅ I've reviewed the code. Looking good! I left a few questions and comments about some behaviors we may want to double-check.

Update: I also ran the e2e tests against a JN site, and they all passed. I was about to test the other subscription-related flows, but it may be better to address Kristófer's comments first.

reykjalin commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -559,22 +559,42 @@ public function add_subscription_information_to_intent( $request, $order, $prepa
+				$request['payment_method_options']['card']['mandate_options']['amount']          = $renewal_amount;
+				$request['payment_method_options']['card']['mandate_options']['reference']       = $order->get_id();
+				$request['payment_method_options']['card']['mandate_options']['start_date']      = $renewal_order->get_time( 'start' );
+				$request['payment_method_options']['card']['mandate_options']['supported_types'] = [ 'india' ];
```
How does this work when there are multiple subscriptions in the renewal order? I'm wondering if we need to add a check to see if `amount_type` should be set to `maximum` and the `interval` to `sporadic`, similar to what we do below.

Better yet, maybe a helper function would be better here that we could use in both places?

```php
function create_mandate_options_for_order( $order ): array {
	$mandate_options = [];
	$subscriptions   = wcs_get_subscriptions_for_order( $order );

	$sub_amount = 0;
	foreach ( $subscriptions as $sub ) {
		$sub_amount += WC_Stripe_Helper::get_stripe_amount( $sub->get_total() );
	}

	// Note: we're making an assumption here that we should be creating the
	// mandate options even if the `$sub_amount` is 0, which we might not want
	// to? Not sure at the time of writing.

	// Get the first subscription associated with this order.
	$sub = reset( $subscriptions );

	if ( 1 === count( $subscriptions ) ) {
		$mandate_options['amount_type']    = 'fixed';
		$mandate_options['interval']       = $sub->get_billing_period();
		$mandate_options['interval_count'] = $sub->get_billing_interval();
	} else {
		// If there are multiple subscriptions the amount_type becomes 'maximum' so we can charge anything
		// less than the order total, and the interval is sporadic so we don't have to follow a set interval.
		$mandate_options['amount_type'] = 'maximum';
		$mandate_options['interval']    = 'sporadic';
	}

	$mandate_options['amount']          = $sub_amount;
	$mandate_options['reference']       = $order->get_id();
	$mandate_options['start_date']      = $sub->get_time( 'start' );
	$mandate_options['supported_types'] = [ 'india' ];

	return $mandate_options;
}
```

That function can then be used in both places like `$request['payment_method_options']['card']['mandate_options'] = $this->create_mandate_options_for_order( $order );`.

Mayisha replied:
Moved it to a separate function as suggested.

reykjalin commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -611,9 +635,51 @@ public function add_subscription_information_to_intent( $request, $order, $prepa
+			}
+
+			// Return from the most recent renewal order with a valid mandate. Mandate is created against a payment method
+			// and for a specific amount in Stripe so the payment method and amount should also match to reuse the mandate.
```
> the payment method and amount should also match to reuse the mandate

Payment method, yes, amount, _not necessarily_. Please correct me if I'm wrong, but I think part of what you're trying to guard against here is when a customer signs up to more than 1 subscription with different renewal periods in the same order?

For example:
- Sub A: renews monthly.
- Sub B: renews weekly.

First week of every month will be more expensive because the renewal order will be paying for both subscriptions. If I understand this current section correctly a different mandate will be created for when only Sub B is renewed, and the initial mandate will be used for when Sub A and Sub B renew at the same time.

I think this would create a confusing experience for the customer because they'll get a sign-up message twice for what they may think of as the same subscription.

Stripe solves this by allowing the `amount_type` to be `maximum` and the `interval` to be `sporadic`, meaning for this order of Sub A + Sub B we can set the `amount_type` to `maximum`, the `amount` to `Sub A + Sub B price`, and the `interval` to `sporadic` (in this example we could probably use `weekly`, but that may not be the case for every mix of renewal intervals and more difficult to implement).

Ultimately that means the price may not match _even though the mandate is valid for the order_, so I think — based on the comment — that this section may need to be rethought a bit? Granted, I don't really have the full context of when this code is executed so I may be wrong 🤔

Mayisha replied:
The scenario I considered here was editing and updating the subscription amount from `wp-admin`. As we discussed in slack, this scenario is not supported. So I removed the amount check and kept the payment method check only. [[3476211](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2633/commits/34762119bd4ddbc032c841b6747894625f7505f9)]

Mayisha committed:
add check for payment_method before accessing from $request array

Mayisha committed:
generate mandate_options in a separate function

Mayisha committed:
Merge branch 'develop' into fix/2607-0-sub

Mayisha commented:
From the last round of review, there were 3 issues discovered,

1.  Adding tax and updating the subscription gives an error `payment_intent_mandate_invalid` [[comment](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2633#discussion_r1264521926)]
2. When the store has different taxes for different states the same `payment_intent_mandate_invalid` happens [[comment](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2633#discussion_r1264521926)]
3. I added a comparison between the current order amount and the previous order amount that has a saved mandate. This check was added to support a specific scenario. Suppose a renewal order A has a saved mandate M for amount 1200 and the merchant edits the subscription to change the price to 1500. Now when shopper tries to renew the order, the renewal order B fails with an error `payment_intent_mandate_invalid` because mandate M was created for amount 1200, not 1500. But this amount comparison will be conflicting for another scenario mentioned [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2633#discussion_r1264767126) (multi sub order)

I had a discussion with @reykjalin and figured that we can ignore the tax update scenario (point 1 above) and the subscription amount update scenario (which I was trying to cover with the amount comparison mentioned in point 3 above) because changing the subscription price isn’t supported. The merchant has to either:

- Ask the customer to repurchase the subscription if the price is changed; or
- Create a new subscription and have the customer come back to the store to process the payment.

As mandates are only created for Indian card and INR currency, and India does not have state based different tax rules we can ignore the scenario in point 2 also.

As changing the subscription price and continuing with the same subscription is not supported, the testing instruction I added in the description under the `Testing with a different amount` becomes invalid. I am mentioning that in the description.

cc: @a-danae @reykjalin

broadeye commented:
Ability for customer to upgrade from a free variation to a paid one on the same subscription is a fundemental feature of WC Subscriptions. Being unable to do so would cripple real world merchant business models.

Mayisha commented:
> Ability for customer to upgrade from a free variation to a paid one on the same subscription is a fundemental feature of WC Subscriptions. Being unable to do so would cripple real world merchant business models.

The functionality for customer's to upgrade from a free to paid option is not changed here. It will still work the same as before.

a-danae commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -560,31 +560,86 @@ public function add_subscription_information_to_intent( $request, $order, $prepa
+	 * @param WC_Order $order The renewal order.
+	 * @return array the mandate_options for the subscription order.
+	 */
+	public function create_mandate_options_for_order( $order ) {
```
Would it be good to have this method as private? We're only using it in this same class.

a-danae commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -666,19 +614,66 @@ public function get_mandate_from_previous_renewal( $order, $payment_method ) {
+	 */
+	public function create_mandate_options_for_order( $order ) {
+		$mandate_options = [];
+		$subscriptions   = wcs_get_subscriptions_for_renewal_order( $order );
```
Probably a nit, but how about passing the subscriptions as a parameter of this `create_mandate_options_for_order` method because we're already retrieving them in line 567, before calling this method?

`wcs_get_subscriptions_for_renewal_order` might not be expensive, though. Not a blocker on my end, I'm fine either way.

a-danae commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -666,19 +614,66 @@ public function get_mandate_from_previous_renewal( $order, $payment_method ) {
+			$mandate_options['interval']    = 'sporadic';
+		}
+
+		$mandate_options['amount']          = $sub_amount;
```
Before, when the subscriptions came from `wcs_get_subscriptions_for_renewal_order`, we used the total of the first subscription...

https://github.com/woocommerce/woocommerce-gateway-stripe/commit/13da844d36c7e9174342613ec70df38ab7c5b7d9#diff-516e329017ff60072191226cd6dfc80b4b9ca8ccccacd6c151c1598cf4179434L585-L586

... as the mandate's amount property

https://github.com/woocommerce/woocommerce-gateway-stripe/commit/13da844d36c7e9174342613ec70df38ab7c5b7d9#diff-516e329017ff60072191226cd6dfc80b4b9ca8ccccacd6c151c1598cf4179434R591 .

Now we're using the sum of all subscriptions. Would that alter the behavior after this refactoring?

Mayisha replied:
Good catch @a-danae. Actually, the logic I was using before ( where we used the total of the first subscription) was faulty. I did not consider a renewal order with multiple subscriptions there which I realized upon getting https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2633#discussion_r1264765793.

a-danae commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -666,19 +614,66 @@ public function get_mandate_from_previous_renewal( $order, $payment_method ) {
-				} else {
-					return '';
-				}
+				return $mandate;
```
Since we aren't comparing the order amount anymore, should we remove lines 610:614?
https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2633/commits/34762119bd4ddbc032c841b6747894625f7505f9#diff-516e329017ff60072191226cd6dfc80b4b9ca8ccccacd6c151c1598cf4179434R610-R614

We already stopped setting the `_stripe_mandate_created_for_amount` meta in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2633/commits/34762119bd4ddbc032c841b6747894625f7505f9#diff-516e329017ff60072191226cd6dfc80b4b9ca8ccccacd6c151c1598cf4179434L638

And should we also remove line 607?
https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2633/commits/34762119bd4ddbc032c841b6747894625f7505f9#diff-516e329017ff60072191226cd6dfc80b4b9ca8ccccacd6c151c1598cf4179434R607

a-danae commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -560,31 +560,86 @@ public function add_subscription_information_to_intent( $request, $order, $prepa
+			if ( $parent_order_id === $renewal_order_id ) {
+				$parent_order         = wc_get_order( $parent_order_id );
+				$renewal_order_amount = $parent_order->get_meta( '_stripe_mandate_created_for_amount', true );
+			}
```
Since we're not comparing the order amount anymore, should we remove some now unused variables?

Thinking of `$renewal_order_amount`, `$parent_order`, `$parent_order_id`, `$order_amount`, and `$parent_order_id`.

a-danae reviewed (commented):
Loving how clean it feels with the mandate properties being handled in a separate method 🙌 

I only reviewed the code, I still need to test it. I left some comments but they're not blockers.

Mayisha committed:
Merge branch 'develop' into fix/2607-0-sub

Mayisha committed:
clean up

a-danae reviewed (commented):
I only noticed one behavior we might want to double-check, but other than that, this looks good to me! Thanks @Mayisha !

✅ Testing saved card with UPE enabled
✅  Testing 0 amount subscription with UPE enabled
✅ Testing saved card with UPE disabled

❓Testing 0 amount subscription with UPE disabled
 I get this error when following the steps:

https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/35834a98-6012-48db-9252-5a6ec85ed7b1


```
Error: stdClass Object
(
    [error] => stdClass Object
        (
            [message] => Mandate or subscription exists with given reference: 262
            [request_log_url] => https://dashboard.stripe.com/test/logs/req_vnyH6S5E94pPqB?t=1701725656
            [type] => invalid_request_error
        )

)
```

Mayisha committed:
Merge branch 'develop' into fix/2607-0-sub

Mayisha commented:
Thanks @a-danae for testing this again. 

I could not reproduce the error that you are having for '0 amount subscription with UPE disabled'. @james-allan also had a look and could not reproduce it either. I have created a new issue for this bug https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2800 and planning to handle it separately as it is not always reproducible.

Mayisha committed:
remove lint file from commit

a-danae reviewed (approved):
Things look good to me 👍 Thanks for double-checking it and for creating an issue 🙂

Glad to approve this since the scenario is very specific and we couldn't replicate this behavior consistently. I'll dive further into it. Thanks again!
